packages/next/src/build/webpack/resolve-plugins/next-resolve-plugin.ts

```diff
@@ -0,0 +1,294 @@
+            let packageName = packageMatch[0] || ''
+
+            switch (packageName) {
+              case 'react-dom':
+              case 'scheduler': {
+                if (this.profilingReact) {
+                  // We are in profiling mode and need to swap in the profiler tracing build
+                  let alias: string
+                  if (requestString === 'react-dom') {
+                    alias = 'react-dom/profiling'
+                  } else if (requestString.startsWith('scheduler/tracing')) {
+                    alias =
+                      'scheduler/tracing-profiling' +
+                      requestString.slice('scheduler/tracing'.length)
+                  } else {
+                    return callback()
+                  }
+
+                  return resolver.doResolve(
+                    target,
+                    {
+                      ...request,
+                      request: alias,
+                    },
+                    'aliasing ' + packageName,
+                    resolveContext,
+                    callback
+                  )
+                }
+                break
+              }
+              default:
+            }
+          }
+
+          return callback()
+        }
+      )
+  }
+}
+
+type NextAppResolvePluginOptions = {
+  experimentalReact?: boolean
+  serverRuntime?: boolean
+}
+
+export class NextAppResolvePlugin {
+  private readonly options: NextAppResolvePluginOptions
+  private readonly serverRuntime: boolean
+  private readonly aliases: Map<string, string>
+  private readonly reactVendoredPackageSuffix: string
+  private readonly vendorPath: string
+
+  constructor(options?: NextAppResolvePluginOptions) {
+    if (!registeredPlugins.has(NextAppResolvePlugin)) {
+      webpack.util.serialization.register(
+        NextAppResolvePlugin,
+        'next/dist/build/webpack/plugins/next-resolve-plugin',
+        'NextAppResolvePlugin',
+        new (class NextAppResolvePluginSerializer {
+          serialize(source: NextAppResolvePlugin, { write }: any) {
+            const o = source.getOptions()
+            write(o)
+          }
+
+          deserialize({ read }: any) {
+            const o = read()
+            return new NextAppResolvePlugin(o)
+          }
+        })()
+      )
+      registeredPlugins.add(NextAppResolvePlugin)
+    }
+
+    this.options = options || {}
+
+    // Some logic below depends on whether we are bundling to run on edge/node vs the browser.
+    this.serverRuntime = (options && options.serverRuntime) || false
+
+    // Alias resolving is bespoke so this is better modeled by a map than with switch branches like we do
+    // with vendored resolves where the logic follows a consistent pattern for most packages. Think of aliasing as
+    // being for individual files and vendoring for entire packages.
+    // This aliasing is subtley different from builtin webpack aliasing. It matches exact request strings rather than path segments.
```

Comment:
```suggestion
    // This aliasing is subtly different from builtin webpack aliasing. It matches exact request strings rather than path segments.
```